source/Shop.h

```diff
@@ -0,0 +1,146 @@
+/* Shop.h
+Copyright (c) 2025 by Amazinite
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "ConditionSet.h"
+#include "DataNode.h"
+#include "LocationFilter.h"
+#include "Sale.h"
+#include "Set.h"
+
+#include <string>
+
+class ConditionsStore;
+class Planet;
+
+
+
+// Class representing a shop of items. Shops are able to be added to planets to designate
+// that this shop should always stock that planet, or they can be given a condition set and/or
+// location filter that allows them to optionally appear on a planet.
+template <class Item>
+class Shop {
+public:
+	Shop();
+	Shop(const DataNode &node, const Set<Item> &items, const ConditionsStore *playerConditions);
+
+	void Load(const DataNode &node, const Set<Item> &items, const ConditionsStore *playerConditions);
+
+	// This shop's name.
+	const std::string &Name() const;
+
+	// All the items that this shop has in stock.
+	const Sale<Item> &Stock() const;
+
+	// Whether this shop is able to stock the given planet.
+	bool CanStock(const Planet *planet) const;
+
+
+private:
+	std::string name;
+	Sale<Item> stock;
+
+	ConditionSet toSell;
+	LocationFilter location;
+};
+
+
+
+template <class Item>
+Shop<Item>::Shop()
+{
+}
+
+
+
+template <class Item>
+Shop<Item>::Shop(const DataNode &node, const Set<Item> &items, const ConditionsStore *playerConditions)
+{
+	Load(node, items, playerConditions);
+}
+
+
+
+template <class Item>
+void Shop<Item>::Load(const DataNode &node, const Set<Item> &items, const ConditionsStore *playerConditions)
+{
+	name = node.Token(1);
+
+	for(const DataNode &child : node)
+	{
+		const bool add = (child.Token(0) == "add");
+		const bool remove = (child.Token(0) == "remove");
+
+		const std::string &key = child.Token((add || remove) ? 1 : 0);
+		const int valueIndex = (add || remove) ? 2 : 1;
+		const bool hasValue = child.Size() > valueIndex;
+
+		if(key == "to" && hasValue && child.Token(valueIndex) == "sell")
+		{
+			if(remove)
+				toSell = ConditionSet{};
+			else
+				toSell.Load(child, playerConditions);
+		}
+		else if(key == "location")
+		{
+			if(remove)
+				location = LocationFilter{};
+			else
+				location.Load(child);
```

Comment:
Same as for conditions; this should not overwrite the existing filters, but merge them. (I don't know if that is supported yet, for ConditionSets I know it is somewhat supported by adding an extra toplevel node.)

Reply:
Ditto as with condition sets. This is how other classes handle events changing location filters.